lib/fastlane/plugin/firebase_app_distribution/helper/firebase_app_distribution_auth_client.rb

```diff
@@ -5,55 +5,73 @@ module Auth
+      #                        CI environment
+      #
+      # env variables
+      #   GOOGLE_APPLICATION_CREDENTIALS - see google_service_path
+      #   FIREBASE_TOKEN - see firebase_cli_token
+      #
+      # Crashes if given invalid or missing credentials
+      def fetch_auth_token(google_service_path, firebase_cli_token)
         if !google_service_path.nil? && !google_service_path.empty?
-          service_account(google_service_path)
-        elsif ENV["FIREBASE_TOKEN"]
-          firebase_token(ENV["FIREBASE_TOKEN"])
-        elsif ENV["GOOGLE_APPLICATION_CREDENTIALS"]
-          service_account(ENV["GOOGLE_APPLICATION_CREDENTIALS"])
+          UI.message("Authenticating with --service_credentials_file path parameter")
+          token = service_account(google_service_path)
+        elsif !firebase_cli_token.nil? && !firebase_cli_token.empty?
+          UI.message("Authenticating with --firebase_cli_token parameter")
+          token = firebase_token(firebase_cli_token)
+        elsif !ENV["FIREBASE_TOKEN"].nil? && !ENV["FIREBASE_TOKEN"].empty?
+          UI.message("Authenticating with FIREBASE_TOKEN environment variable")
+          token = firebase_token(ENV["FIREBASE_TOKEN"])
+        elsif !ENV["GOOGLE_APPLICATION_CREDENTIALS"].nil? && !ENV["GOOGLE_APPLICATION_CREDENTIALS"].empty?
+          UI.message("Authenticating with GOOGLE_APPLICATION_CREDENTIALS environment variable")
+          token = service_account(ENV["GOOGLE_APPLICATION_CREDENTIALS"])
         elsif (refresh_token = refresh_token_from_firebase_tools)
-          firebase_token(refresh_token)
+          UI.message("Authenticating with the firebase login:ci command")
+          token = firebase_token(refresh_token)
         else
-          UI.crash!(ErrorMessage::MISSING_CREDENTIALS)
+          UI.user_error!(ErrorMessage::MISSING_CREDENTIALS)
         end
+        UI.success("Successfully authenticated!")
+        token
       end
 
       private
 
       def refresh_token_from_firebase_tools
-        if ENV["XDG_CONFIG_HOME"]
-          config_path = File.expand_path("configstore/firebase-tools.json", ENV["XDG_CONFIG_HOME"])
-        else
+        if ENV["XDG_CONFIG_HOME"].nil? || ENV["XDG_CONFIG_HOME"].empty?
           config_path = File.expand_path(".config/configstore/firebase-tools.json", "~")
+        else
+          config_path = File.expand_path("configstore/firebase-tools.json", ENV["XDG_CONFIG_HOME"])
         end
 
         if File.exist?(config_path)
           begin
-           refresh_token = JSON.parse(File.read(config_path))['tokens']['refresh_token']
-           unless refresh_token.nil? && refresh_token.empty?
-             refresh_token
-           end
-         # Rescue is empty to return nil instead of an error
-         # when there is an empty "tokens" field in the firebase-tools json
-         rescue NoMethodError
-         end
+            refresh_token = JSON.parse(File.read(config_path))['tokens']['refresh_token']
```

Comment:
I don't think we should address this in this particular CL, but I _think_ we want to catch and swallow any `JSON::ParserError` that `Json.parse` may raise. Also, feels a little scary rescuing a generic `NoMethodError` here. We should probably use some nested if statements. Not as pretty, but less likely to swallow errors that we shouldn't be swallowing.

Reply:
good point, I'll leave a comment in the code for this